[apollo-client] Fix issue with query/mutate functions and passing variables

ganemone opened the pull request:
- Links to documentation: https://www.apollographql.com/docs/react/api/apollo-client#ApolloClient.query
- Link to GitHub or NPM: https://github.com/apollographql/apollo-client/
- Type of contribution: fix

This fixes an issue where variables could not be passed to query/mutate functions and adds a regression test.

Fixes #2888

indeyets commented:
@ganemone add `[apollo-client]` to the title please

ganemone commented:
> We generally avoid using `any` where possible, and this fix loses type information. E.g. we do know the `variables` option would be an object.
> 
> Can you work toward a solution that addresses the problem without dropping knowledge of it being an object?

That is fair. There is a solution which would let us not use `any` however it would require significant changes that would be breaking for many users as it would moving many of the exports to be type aliases rather than interfaces. I'm happy to do that but I thought I would introduce the less intrusive option first.

AndrewSouthpaw commented:
Sure, that's understandable. Generally we've suggested moving away from type interfaces when you can get away with it, because they've often triggered more slowness in flow.

Do you suspect that many people have actually been using that exported type as an interface? How bad of a breaking change would it be?

ganemone commented:
Now that I think about it, it might not be that big of a breaking change since it was broken when passing variables in the first place. And it should be non breaking unless they are using features like `extends`.

I updated the diff to move the affected exports from interfaces to types and removed usages of `any`.

AndrewSouthpaw commented on definitions/npm/apollo-client_v2.x.x/flow_v0.57.x-/apollo-client_v2.x.x.js:
```diff
@@ -495,7 +494,7 @@ declare module "apollo-client" {
   declare interface GraphQLRequest {
     query: DocumentNode;
-    variables?: { [key: string]: any };
+    variables?: any;
```
Change this line back?

AndrewSouthpaw commented on definitions/npm/apollo-client_v2.x.x/flow_v0.57.x-/apollo-client_v2.x.x.js:
```diff
@@ -278,8 +278,7 @@ declare module "apollo-client" {
 
-  declare export interface WatchQueryOptions
-    extends ModifiableWatchQueryOptions {
+  declare export type WatchQueryOptions = ModifiableWatchQueryOptions & {
```
Generally I subtype by doing:

```js
type Subtype = {
  ...$Exact<Supertype>,
  foo: string
}
```

There are times when doing an intersection creates weird outcomes.

ganemone commented:
Any further comments?

AndrewSouthpaw commented:
Thanks!
